### PR TITLE
move caret with mouse click in editbox

### PR DESCRIPTION
**Pull request by jmoerschbach**
move cursor position by clicking with the mouse in an editbox

### CODE REVIEW

File: src/ui_basic/editbox.cc [hunk 1]
```diff
@@ -212,6 +213,17 @@ bool EditBox::handle_mousepress(const uint8_t btn, int32_t, int32_t) {
 	return false;
 }
 
+void EditBox::set_caret_to_cursor_pos(int32_t x) {
+	int textWidth = text_width(m_->text, *m_->font_style, m_->font_scale);
```
Comment (Noordfrees) [inline src/ui_basic/editbox.cc]: please use under_scores instead of CamelCase

**Review by Noordfrees — CHANGES_REQUESTED** [review]
This doesn't quite work when the letters have very uneven width. Try typing "iiiiiiiimmmmmmmm" in an edit box and then clicking – the cursor jumps to several letters away from the clicked spot.

---

```
[193/332] Building CXX object src/ui_basic/CMakeFiles/ui_basic.dir/editbox.cc.o
../src/ui_basic/editbox.cc: In member function ‘virtual bool UI::EditBox::handle_mousepress(uint8_t, int32_t, int32_t)’:
../src/ui_basic/editbox.cc:205:71: warning: unused parameter ‘y’ [-Wunused-parameter]
  205 | bool EditBox::handle_mousepress(const uint8_t btn, int32_t x, int32_t y) {
      |                                                               ~~~~~~~~^
```

**Comment by klaus-halfmann** [comment]
@Noordfrees is corrct this will not work as charatces have uneven spacing.
You will need a binay search approxmation findig the best location.
@jmoerschbach I am not _that_ familiar with the font library, but if you give me some hints
I could try to implement this.

OTOH we desperaterly nee this feature, thanks for starting this.

**Comment by jmoerschbach** [comment]
feels not very elegant (and probably breaks with rtl) but works way better now :)

**Review by Noordfrees — COMMENTED** [review]
Code LGTM, tested and works well with almost all strings.
But…
Widelands (for whatever reason) ignores characters äöü etc on my machine. I can enter them into an edit box by typing them in a text editor, copying them, and pasting into the edit box. When I then click into the string `äöü`, the cursor jumps to a place *in the middle of* a character (where it can not be moved by just using the arrow keys). Pressing backspace then either segfaults (release build so no backtrace, sorry) or changes the text in the edit box from grafik to grafik.

File: src/ui_basic/editbox.cc [hunk 2]
```diff
@@ -212,6 +213,55 @@ bool EditBox::handle_mousepress(const uint8_t btn, int32_t, int32_t) {
 	return false;
 }
 
+void EditBox::set_caret_to_cursor_pos(int32_t x) {
+	if (x <= 2 * kMarginX) {
```
Comment (klaus-halfmann) [inline src/ui_basic/editbox.cc]: when / how can this happen?

File: src/ui_basic/editbox.cc [hunk 3]
```diff
@@ -212,6 +213,55 @@ bool EditBox::handle_mousepress(const uint8_t btn, int32_t, int32_t) {
...
+		set_caret_pos(0);
+		return;
+	}
+
+	int text_w = text_width(m_->text, *m_->font_style, m_->font_scale);
+	if (text_w <= 0) {
```
Comment (klaus-halfmann) [inline src/ui_basic/editbox.cc]: // no visible text

File: src/ui_basic/editbox.cc [hunk 4]
```diff
@@ -212,6 +213,55 @@ bool EditBox::handle_mousepress(const uint8_t btn, int32_t, int32_t) {
...
+
+	// mouse coordinate x=0 means leftmost spot in editbox but text starts with margin -> adjust
+	x -= kMarginX;
+	text_w = calculate_text_width(index);
+
+	index = approximate_cursor(x, text_w, index);
+
+	set_caret_pos(index);
+}
+int EditBox::approximate_cursor(int32_t x, int text_w, int index) const {
```
Comment (klaus-halfmann) [inline src/ui_basic/editbox.cc]: This is no binary search but an incremental approximation, but it should terminate in any case.
Any idea how often the loops has to be run?

**Comment by klaus-halfmann** [comment]
Welcome to the wolrd of Unicode characters. AFAIK (Using Java only)
The "real" Unicode characters a wchar_t which is perpaps actually 32bit or evene bigger.
e.g. https://www.cprogramming.com/tutorial/unicode.html
So when dealing with UTF you may use Encondings like UTF8, UTF16 (wich Java does) or the "real UTF" / wchar_t

Problem with UTF8 and (less ifeten) UTF16 is that a single byte may:
1) not identify a single character
2) enconding of some characters is not unique

e.g. the German Umlauts ÄÖÜ _may_ be econde aloing some ISO Encoding, _or_ encode as (prefix") + A etc.

Thus said when you tear apart such an multibyte enconded character you may end up with gargabe chracters
or even crashing (the std. unicode  liraries has some nasty security vulnerabilities as of this).

Ill take a look at you code now and consult some docs. I hope @gunchleoc can shed some more light on this.

File: src/ui_basic/editbox.h [hunk 5]
```diff
@@ -104,6 +104,9 @@ struct EditBox : public Panel {
 	void calculate_selection_boundaries(uint32_t& start, uint32_t& end);
 	void delete_selected_text();
 	void copy_selected_text();
+	void set_caret_to_cursor_pos(int32_t x);
```
Comment (klaus-halfmann) [inline src/ui_basic/editbox.h]: Plase add some comments, epecially about the context of x, index and text_w

**Review by klaus-halfmann — CHANGES_REQUESTED** [review]
As of @Noordfrees and my commetns this needs some improvement.

I will compile this now and try, macOS normally should handle umlauts quite easily,
lets see.

File: src/ui_basic/editbox.cc (hunk 4 above)
Comment (jmoerschbach) [reply src/ui_basic/editbox.cc]: this depends on string length and char distribution of course. In my tests for input string "iiiiiiiiiimmmmmmmmmmmiiiiiii" ("worst case" regarding char width distribution) it was about 10 loops. With real world strings, the very first approximation is already quite good so the loop only needs to run ~2-3 times :)

File: src/ui_basic/editbox.cc (hunk 2 above)
Comment (jmoerschbach) [reply src/ui_basic/editbox.cc]: this happens when you click on the leftmost spot (left border) of the widget. x=0 is not the start of the first char (there is a margin before the text begins).

**Comment by jmoerschbach** [comment]
should work now :) (really this time^^)

File: src/ui_basic/editbox.h (hunk 5 above)
Comment (jmoerschbach) [reply src/ui_basic/editbox.h]: changed the names. descriptive code > comments :)

File: src/ui_basic/editbox.cc (hunk 3 above)
Comment (jmoerschbach) [reply src/ui_basic/editbox.cc]: improved by merging the two boundary checks into one. should read good enough without any additional comments

**Comment by Noordfrees** [comment]
ÄÖÜ work fine now, but the word `HuſumðÐ` still shows the same error after clicking on the 'ð'.
Now I used a debug build so I have a backtrace for the segfault:
```
AddressSanitizer:DEADLYSIGNAL
=================================================================
==8959==ERROR: AddressSanitizer: SEGV on unknown address 0x6061002137ff (pc 0x562ac036b7c3 bp 0x7ffc1ba50650 sp 0x7ffc1ba50460 T0)
==8959==The signal is caused by a READ memory access.
    #0 0x562ac036b7c3 in UI::EditBox::handle_key(bool, SDL_Keysym) ../src/ui_basic/editbox.cc:355
    #1 0x562ac03a9c66 in UI::Panel::do_key(bool, SDL_Keysym) ../src/ui_basic/panel.cc:1147
    #2 0x562ac03a9bb0 in UI::Panel::do_key(bool, SDL_Keysym) ../src/ui_basic/panel.cc:1139
    #3 0x562ac03a9bb0 in UI::Panel::do_key(bool, SDL_Keysym) ../src/ui_basic/panel.cc:1139
    #4 0x562ac03a9bb0 in UI::Panel::do_key(bool, SDL_Keysym) ../src/ui_basic/panel.cc:1139
    #5 0x562ac03a9bb0 in UI::Panel::do_key(bool, SDL_Keysym) ../src/ui_basic/panel.cc:1139
    #6 0x562ac03a9bb0 in UI::Panel::do_key(bool, SDL_Keysym) ../src/ui_basic/panel.cc:1139
    #7 0x562ac03a9bb0 in UI::Panel::do_key(bool, SDL_Keysym) ../src/ui_basic/panel.cc:1139
    #8 0x562ac03a9bb0 in UI::Panel::do_key(bool, SDL_Keysym) ../src/ui_basic/panel.cc:1139
    #9 0x562ac03a9bb0 in UI::Panel::do_key(bool, SDL_Keysym) ../src/ui_basic/panel.cc:1139
    #10 0x562ac03a9bb0 in UI::Panel::do_key(bool, SDL_Keysym) ../src/ui_basic/panel.cc:1139
    #11 0x562ac03a9bb0 in UI::Panel::do_key(bool, SDL_Keysym) ../src/ui_basic/panel.cc:1139
    #12 0x562ac03ab281 in UI::Panel::ui_key(bool, SDL_Keysym) ../src/ui_basic/panel.cc:1371
    #13 0x562abfd54949 in WLApplication::handle_input(InputCallback const*) ../src/wlapplication.cc:689
    #14 0x562ac03a12d9 in UI::Panel::do_run() ../src/ui_basic/panel.cc:216
    #15 0x562abfd79e29 in int UI::Panel::run<int>() (/home/benedikt/wl/enhance_launch_editor/widelands+0x51de29)
    #16 0x562abfd517e1 in WLApplication::run() ../src/wlapplication.cc:563
    #17 0x562abfd49e51 in main ../src/main.cc:47
    #18 0x7ff18629acb1 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x28cb1)
    #19 0x562abfd49ced in _start (/home/benedikt/wl/enhance_launch_editor/widelands+0x4edced)

AddressSanitizer can not provide additional info.
SUMMARY: AddressSanitizer: SEGV ../src/ui_basic/editbox.cc:355 in UI::EditBox::handle_key(bool, SDL_Keysym)
==8959==ABORTING
```
Version 7002536@HEAD

**Comment by jmoerschbach** [comment]
sry -.-
next try... :D

**Review by Noordfrees — APPROVED** [review]
Works now :)

File: src/ui_basic/editbox.cc [hunk 6]
```diff
@@ -212,6 +213,54 @@ bool EditBox::handle_mousepress(const uint8_t btn, int32_t, int32_t) {
...
+		return;
+	}
+
+	int text_w = text_width(m_->text, *m_->font_style, m_->font_scale);
+
+	// mouse coordinate cursor_pos_x=0 means leftmost spot in editbox but text starts with margin ->
+	// adjust
+	cursor_pos_x -= kMarginX;
+
+	double x_relative = static_cast<double>(cursor_pos_x - m_->scrolloffset) / text_w;
```
Comment (klaus-halfmann) [inline src/ui_basic/editbox.cc:221]: float or even some integer arithmethis would be enoughm. sorrry nitpicking

**Review by klaus-halfmann — APPROVED** [review]
Once Upon a time I qill make this a real binary search but for now its good enoug.
Used google tranlated to fetch soem russian, tradditional chinises, Japanese, Mongoligna an arabic chracters. no crashed. Cursor appeared whre I wanted it to be.

**Pull request merged**